extensions/aws/s3/S3Wrapper.h

```diff
@@ -31,11 +47,190 @@ namespace minifi {
+};
+
+struct ListRequestParameters {
+  std::string bucket;
+  std::string delimiter;
+  std::string prefix;
+  bool use_versions = false;
+  uint64_t min_object_age = 0;
+};
+
+struct ListedObjectAttributes {
+  std::string filename;
+  std::string etag;
+  bool is_latest = false;
+  int64_t last_modified = 0;
+  int length = 0;
+  std::string store_class;
+  std::string version;
+};
+
+using HeadObjectRequestParameters = GetObjectRequestParameters;
+
+class S3Wrapper {
+ public:
+  S3Wrapper();
+  explicit S3Wrapper(std::unique_ptr<S3RequestSender> request_sender);
+
+  void setCredentials(const Aws::Auth::AWSCredentials& cred);
+  void setRegion(const Aws::String& region);
+  void setTimeout(uint64_t timeout);
+  void setEndpointOverrideUrl(const Aws::String& url);
+  void setProxy(const ProxyOptions& proxy);
+
+  minifi::utils::optional<PutObjectResult> putObject(const PutObjectRequestParameters& options, std::shared_ptr<Aws::IOStream> data_stream);
+  bool deleteObject(const std::string& bucket, const std::string& object_key, const std::string& version = "");
+  minifi::utils::optional<GetObjectResult> getObject(const GetObjectRequestParameters& get_object_params, const std::shared_ptr<io::BaseStream>& fetched_body);
+  minifi::utils::optional<std::vector<ListedObjectAttributes>> listBucket(const ListRequestParameters& params);
+  minifi::utils::optional<std::map<std::string, std::string>> getObjectTags(const std::string& bucket, const std::string& object_key, const std::string& version = 0);
+  minifi::utils::optional<HeadObjectResult> headObject(const HeadObjectRequestParameters& head_object_params);
+
+  virtual ~S3Wrapper() = default;
+
  protected:
-  minifi::utils::optional<Aws::S3::Model::PutObjectResult> sendPutObjectRequest(const Aws::S3::Model::PutObjectRequest& request) override;
-  bool sendDeleteObjectRequest(const Aws::S3::Model::DeleteObjectRequest& request) override;
-  minifi::utils::optional<Aws::S3::Model::GetObjectResult> sendGetObjectRequest(const Aws::S3::Model::GetObjectRequest& request) override;
+  void setCannedAcl(Aws::S3::Model::PutObjectRequest& request, const std::string& canned_acl) const;
+  int64_t writeFetchedBody(Aws::IOStream& source, const int64_t data_size, const std::shared_ptr<io::BaseStream>& output);
+  Expiration getExpiration(const std::string& expiration);
+  static std::string getEncryptionString(Aws::S3::Model::ServerSideEncryption encryption);
+
+  minifi::utils::optional<std::vector<ListedObjectAttributes>> listVersions(const ListRequestParameters& params);
+  minifi::utils::optional<std::vector<ListedObjectAttributes>> listObjects(const ListRequestParameters& params);
+  void addListResults(const Aws::Vector<Aws::S3::Model::ObjectVersion>& content, const uint64_t min_object_age, std::vector<ListedObjectAttributes>& listed_objects);
+  void addListResults(const Aws::Vector<Aws::S3::Model::Object>& content, const uint64_t min_object_age, std::vector<ListedObjectAttributes>& listed_objects);
```

Comment:
the top level `const` on the parameters only has an effect in the definition

Reply:
Could you elaborate on this please? I'm not sure I understand the reason behind this comment.

Reply:
the `const` in the parameter `const uint64_t min_object_age` is an "implementation detail" https://abseil.io/tips/109

Reply:
Thanks for the link, I never thought about it from this point of view, but it makes sense to follow the mentioned guideline.

Reply:
Updated in [4507f26](https://github.com/apache/nifi-minifi-cpp/pull/975/commits/4507f26caf3e1bec015f411fc1e924a08fa2c9a3)